common/quotas/priority_rate_limiter_impl.go

```diff
@@ -0,0 +1,163 @@
+	apiToPriority map[string]int,
+	priorityToRateLimiters map[int]RateLimiter,
+) *PriorityRateLimiterImpl {
+	priorities := make([]int, 0, len(priorityToRateLimiters))
+	for priority := range priorityToRateLimiters {
+		priorities = append(priorities, priority)
+	}
+	sort.Slice(priorities, func(i, j int) bool {
+		return priorities[i] < priorities[j]
+	})
+	priorityToIndex := make(map[int]int, len(priorityToRateLimiters))
+	rateLimiters := make([]RateLimiter, 0, len(priorityToRateLimiters))
+	for index, priority := range priorities {
+		priorityToIndex[priority] = index
+		rateLimiters = append(rateLimiters, priorityToRateLimiters[priority])
+	}
+
+	return &PriorityRateLimiterImpl{
+		apiToPriority:          apiToPriority,
+		priorityToRateLimiters: priorityToRateLimiters,
+
+		priorityToIndex: priorityToIndex,
+		rateLimiters:    rateLimiters,
+	}
+}
+
+func (p *PriorityRateLimiterImpl) Allow(
+	now time.Time,
+	request Request,
+) bool {
+	decidingRateLimiter, consumeRateLimiters := p.getRateLimiters(request)
+
+	allow := decidingRateLimiter.AllowN(now, request.Token)
+	if !allow {
+		return false
+	}
+
+	for _, limiter := range consumeRateLimiters {
+		_ = limiter.ReserveN(now, request.Token)
+	}
+	return allow
+}
+
+func (p *PriorityRateLimiterImpl) Reserve(
+	now time.Time,
+	request Request,
+) Reservation {
+	decidingRateLimiter, consumeRateLimiters := p.getRateLimiters(request)
+
+	decidingReservation := decidingRateLimiter.ReserveN(now, request.Token)
+	if !decidingReservation.OK() {
+		return decidingReservation
+	}
+
+	otherReservations := make([]Reservation, len(consumeRateLimiters))
+	for index, limiter := range consumeRateLimiters {
+		otherReservations[index] = limiter.ReserveN(now, request.Token)
+	}
+	return NewPriorityReservation(decidingReservation, otherReservations)
+}
+
+func (p *PriorityRateLimiterImpl) Wait(
+	ctx context.Context,
+	request Request,
+) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	now := time.Now().UTC()
+	reservation := p.Reserve(now, request)
+	if !reservation.OK() {
+		return fmt.Errorf("rate: Wait(n=%d) would exceed context deadline", request.Token)
+	}
+
+	delay := reservation.DelayFrom(now)
+	if delay == 0 {
+		return nil
+	}
+	waitLimit := InfDuration
+	if deadline, ok := ctx.Deadline(); ok {
+		waitLimit = deadline.Sub(now)
+	}
+	if waitLimit < delay {
+		reservation.CancelAt(now)
+		return fmt.Errorf("rate: Wait(n=%d) would exceed context deadline", request.Token)
+	}
+
+	t := time.NewTimer(delay)
+	defer t.Stop()
+	select {
+	case <-t.C:
+		return nil
+
+	case <-ctx.Done():
+		reservation.CancelAt(time.Now())
+		return ctx.Err()
+	}
+}
+
+func (p *PriorityRateLimiterImpl) getRateLimiters(
+	request Request,
+) (RateLimiter, []RateLimiter) {
+	priority, ok := p.apiToPriority[request.API]
+	if !ok {
+		// if API not assigned a priority use the lowest priority
+		return p.rateLimiters[len(p.rateLimiters)-1], nil
+	}
+
+	rateLimiterIndex := p.priorityToIndex[priority]
```

Comment:
what if this priority does not exist in the priorityToIndex (e.g. it has a weird priority like 112358)?  It seems we will assume it as a high priority call?

Reply:
added a check in initialization